File: 0101. Symmetric Tree/101.symmetric-tree.go
package leetcode

/*
 * @lc app=leetcode id=101 lang=golang
 *
 * [101] Symmetric Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	// Method 1. DFS
	// if root == nil {
	// 	return true
	// }
	// return isMirror(root.Left, root.Right)

	// Method 2. BFS
	queue := make([][2]*TreeNode, 1)
	queue = append(queue, [2]*TreeNode{root.Left, root.Right})
	for len(queue) > 0 {
		leftNode := queue[0][0]
		rightNode := queue[0][1]
		queue = queue[1:]

		if leftNode == nil && rightNode == nil {
			continue
		}
		if leftNode == nil || rightNode == nil {
			return false
		}
		if leftNode.Val != rightNode.Val {
			return false
		}

		queue = append(queue, [2]*TreeNode{leftNode.Left, rightNode.Right})
		queue = append(queue, [2]*TreeNode{rightNode.Left, leftNode.Right})
	}
	return true
}

// DFS
// func isMirror(n1 *TreeNode, n2 *TreeNode) bool {
// 	if n1 == nil && n2 == nil {
// 		return true
// 	}
// 	if n1 == nil || n2 == nil {
// 		return false
// 	}
// 	return (n1.Val == n2.Val && isMirror(n1.Left, n2.Right) && isMirror(n1.Right, n2.Left))
// }

// @lc code=end

File: 0101. Symmetric Tree/101.symmetric-tree.java
//Given the root of a binary tree, check whether it is a mirror of itself (i.e.,
// symmetric around its center).
//
//
// Example 1:
//
//
//Input: root = [1,2,2,3,4,4,3]
//Output: true
//
//
// Example 2:
//
//
//Input: root = [1,2,2,null,3,null,3]
//Output: false
//
//
//
// Constraints:
//
//
// The number of nodes in the tree is in the range [1, 1000].
// -100 <= Node.val <= 100
//
//
//
//Follow up: Could you solve it both recursively and iteratively?
//
// Related Topics Tree Depth-First Search Breadth-First Search Binary Tree 👍 15
//911 👎 408


//leetcode submit region begin(Prohibit modification and deletion)

import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {
        // Method 1. DFS
        // if (root == null) return true;
        // return isMirror(root.left, root.right);

        // Method 2. BFS
        if (root == null) return true;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root.left);
        queue.add(root.right);

        while (!queue.isEmpty()) {
            TreeNode n1 = queue.poll();
            TreeNode n2 = queue.poll();

            if (n1 == null && n2 == null) continue;
            if (n1 == null || n2 == null || n1.val != n2.val) return false;

            queue.add(n1.left);
            queue.add(n2.right);
            queue.add(n1.right);
            queue.add(n2.left);
        }
        return true;
    }

    private boolean isMirror(TreeNode n1, TreeNode n2) {
        if (n1 == null && n2 == null) return true;
        if (n1 == null || n2 == null) return false;
        return (n1.val == n2.val) && isMirror(n1.left, n2.right) && isMirror(n1.right, n2.left);

    }
}
//leetcode submit region end(Prohibit modification and deletion)

